### DISCUSSION TITLE
Subscribe to reverse Foreign Key changes

### DISCUSSION THREAD

**Discussion by joseluratm** [NilCoalescing/djangochannelsrestframework#121] [Q&A]
First of all, thank you very much for your work. It is a great library!

I am currently facing a casuistica that I do not know if it is because of bad implementation of mine, but I can not find the best solution.

I'll explain my case, to see if you can shed some light on my problem.

Right now I have the following models:

```
class customUser(AbstractUser):
    email = models.EmailField(max_length=50, unique=True)


class event(models.Model):
    title = models.CharField(max_length=200)
    assigned_to = models.ForeignKey(customUser, blank=False, on_delete=models.DO_NOTHING, related_name='assigned_user_case')
```

in my web front end, I can assign each of the events to different users. That is to say, a user can have many events assigned to him and an event can only be assigned to one user.

My goal is that I want to be able to be able to receive all signals from events that bind/unbind to a user **filtering by the user**. This means that only subscription and unsubscription updates per user are sent to each user.

I have tried using the filtering example you specify in the main readme of the project, but I am not able to get it to work. Also, I don't know if that example would be enough to do what I want.

Thank you very much for your help.

**Comment by hishnash** [comment]
Sorry it's taken so long to get back to you.

You will need to use a filtering based on the user pk.

```python
class MyConsumer(AsyncAPIConsumer):

  @model_observer(models.event)
  async def event_change_handler(
      self,
      message,
      observer=None,
      action=None,
      subscribing_request_ids=[],
      **kwargs
  ):
      for request_id in subscribing_request_ids:
          await self.send_json(dict(body=message, action=action, request_id=request_id))

  @event_change_handler.groups_for_signal
  def classroom_change_handler(self, instance: models.event, **kwargs):
      # this block of code is called very often *DO NOT make DB QUERIES HERE*
      yield f'-user-event.{instance.assigned_to_id}'

  @event_change_handler.groups_for_consumer
  def event_change_handler(self, school=None,  user=None, **kwargs):
      # This is called when you subscribe/unsubscribe
     yield   f'-user-event.{user.id}'

  @action()
  async def subscribe_to_events(self, request_id, **kwargs):
      # check user has permission to do this
      await self.event_change_handler.subscribe(user=self.scope['user'], request_id=request_id)
``` 


This will mean when users subscribe they subscribe to all changes to `event` models that point to the user. 

Note if an `event` model is updated to change the old user that as assigned will get a `delete` message and the new user will get a `create` message (since from their perspective the event has ben deleted or created not changed).

**Reply by joseluratm** [reply]
Hello, first of all thank you very much for your reply.

I don't quite understand how it would work.
First of all, I don't see where `-user-event.` comes from.
It is probably a lack of understanding on my part, but does the `-` before each word mean something? In some examples of the documentation at the end of the line it uses `__`, and here you indicate to use the `.` is it for some special reason?

Where it says `-user-event`, in my case it would really be `-customUser-event`, right?

In the way that you indicate and making some changes I have managed to subscribe to all the changes in the events that belong to my user but I still do not receive notification when changing a user event.
Attached is my code in case you could give me a hand.

```
    @event_change.groups_for_consumer
    def event_change(self, user=None, **kwargs):
        if user is not None:
            yield f'-event__{user.id}'


    @event_change.groups_for_signal
    def event_change(self, instance, **kwargs):
        # this block of code is called very often *DO NOT make DB QUERIES HERE*
        yield f'-event__{instance.assigned_to_id}'

    @action()
    async def subscribe_to_events_by_user(self, request_id, **kwargs):
        username = self.scope['user']
        user = customUser.objects.get(username=username)
        await self.event_change.subscribe(user=user, request_id=request_id)

```

Thank you very much!

**Reply by hishnash** [reply]
> It is probably a lack of understanding on my part, but does the - before each word mean something? In some examples of the documentation at the end of the line it uses __, and here you indicate to use the . is it for some special reason?

You can use anything this is just a key you define. This is just something that you partition the events so that when the events happen you can split them up and send the correct messages over the channel groups and each consumer can subscribe only the the groups that they want/have permission to.

> In the way that you indicate and making some changes I have managed to subscribe to all the changes in the events that belong to my user but I still do not receive notification when changing a user event.


Im not sure I understand.  The above code should inform a subscribed consumer about changes that take place for the logged in user.  

Do you want to be informed about changes for other users?

**Reply by joseluratm** [reply]
sorry, I must have explained myself wrong!
What I want is that when the event changes the assigned user, a notification is sent to that same user. Here is an example:
1. The user X is assigned to the event E1.
2. User Y says that this event is his.
3. User X receives that the user assigned to that event has been changed.


Thank you very much!

**Reply by hishnash** [reply]
Do you want user X to know that is was assigned to Y or just that is has now been unassigned from them?

**Reply by joseluratm** [reply]
Notify user X that the event has been unassigned

**Reply by hishnash** [reply]
The current implementation you have there should do this, user X should get notified that the 'event' as been deleted when it is assigned to a new user. Is that not happening for you?

**Reply by joseluratm** [reply]
It is not happening to me. Maybe it's some implementation problem on my part somewhere else in the code.

Anyway, thank you very much!!!

**Reply by hishnash** [reply]
are you able to share how you are changing the 'events' assigned user?  

What should be happening is that the current user gets a message 'delete' (if they are subscribed when this change takes place).  are you able to put a print statement in

```python
  @model_observer(models.event)
  async def event_change_handler(
      self,
      message,
      observer=None,
      action=None,
      subscribing_request_ids=[],
      **kwargs
  ):
      for request_id in subscribing_request_ids:
          await self.send_json(dict(body=message, action=action, request_id=request_id))
```
Bringing out the `message` and `action`
